### ISSUE TITLE
Build out some integration with Cloud Bigtable

### ISSUE THREAD

**Issue by jgeewax** [GoogleCloudPlatform/gcloud-python#872]
We want to make users of Happybase "just work", so...some ideas:

Option 1: A wrapped import module

```python
from gcloud.bigtable import happybase
```

Option 2: Some sort of monkey patching?

```python
from gcloud import bigtable
import happybase
happybase = bigtable.monkey_patch(happybase)
```

**Comment by dhermes** [comment]
Relevant docs:
https://happybase.readthedocs.org/en/latest/
https://cloud.google.com/bigtable/docs/

**Comment by tseaver** [comment]
ISTM that monkey-patching the `happybase` third-party module would be a poor choice.  I think we need to think of this as "write a happybase API emulation layer" instead (which may or may not itself even *use* the real `happybase` code).

**Comment by dhermes** [comment]
Can you elaborate? Why is it a bad idea? (I have yet to look at the `happybase` code/docs.)

**Comment by tseaver** [comment]
I think our goal is that they should be able to port scripts easily (ideally, just changing `import happytable` to `from gcloud.bigtable import happytable` and adjusting the connection parameters).  I don't see interfering with the actual `happytable` module's internals as a useful way to approach that goal.  For instance, users of `gcloud.bigtable` might need to interact with a *real* `happybase`/`Thrift` backend (e.g., to move or compare data between it and the `bigtable` dataset).

**Comment by jgeewax** [comment]
I agree that there may be times where someone would want to talk to Cloud Bigtable *and* HBase at the same time, so monkey-patching happybase sounds like we'd screw over those multi-tenant situations.

Maybe there's a way to `from gcloud import bigtable` and then `happybase_compatible_client = bigtable.get_happybase_client()` which uses the real happybase code, but swaps out a connection of sorts at the lower level...

**Comment by tseaver** [comment]
The gcloud.bigtable docs say:

> Cloud Bigtable works with Hadoop 2.4 and higher and is interface compatible with Apache HBase 1.0.

If that means that it exports the same [REST API](http://hbase.apache.org/book.html#_rest) as HBase / Thrift, then maybe we don't have to do much to enable `happybase`, beyond ensuring that we set the connection up with gcloud-relevant credentials.

**Comment by jgeewax** [comment]
This is the interesting piece. It is not wire-protocol compatible. That is -- it doesn't speak Thrift (AFAIK). Instead, they made the Java code compatible, and swapped out the thing that does serialization and sending of requests underneath. 

I think @maxluebbe can comment more to confirm.

**Comment by dhermes** [comment]
Yeah that one is important. I was under the impression from @jgeewax previous comments that the web APIs were the same / close cousins.

**Comment by brianolson** [comment]
The Java "HBase" client to Bigtable appears to be an emulation layer. A Python client should be built on the same protobuf rpc interface as that:
https://github.com/GoogleCloudPlatform/cloud-bigtable-client/tree/master/bigtable-protos/src/main/proto/google

I'd imagine that Python client having a native mode that just handles the proto-rpc; and then there could be a happybase emulation layer on top of that.

**Comment by maxluebbe** [comment]
Sorry have been OOO. Some clarifications.
1) Additional work item is modernizing Happybase to 1.0 HBase API.
2) Cloud Bigtable does not support Thrift/Rest interfaces. Part of the work that needs to be done here is decoupling the thrift parts of happybase from the interface provided to the user, so that we can drop in gRPC in place of Thrift.

**Comment by jgeewax** [comment]
gRPC or Protobuf over HTTP 1.0?

gRPC is not possible right now given the installation process (we need to keep our install to `pip install gcloud` and can't ask people to do a bunch of Makefile madness to use our library).

**Comment by brianolson** [comment]
Is it acceptable for gRPC to be an optional install? (and python bigtable might not work without it, but the rest of gcloud-python would)
Maybe spin off the python bigtable client into something not part of 'gcloud-python'? (Seems fair, the Java bigtable client is its own project/repo.)

**Comment by jgeewax** [comment]
You're telling me there's no way to talk to Cloud Bigtable's API without using a gRPC client? We can't use HTTP 1.0 ? That seems.... really really weird. @maxluebbe @carterpage can you confirm?

**Comment by carterpage** [comment]
We're designed to be very low latency and high-throughput so all of our focus is on gRPC.  There is a REST interface, which comes for free with our envelope, but we don't recommend using for much more than slow admin API calls or ad hoc test queries against the data API.  Almost everything in the API is a binary string, including keys and values, so there would be a lot of base64 encoding going on -- a big CPU hit on both sides.  We also have not tested streaming scans would work in REST.

**Comment by dhermes** [comment]
Can we help make the gRPC install process easy?

**Comment by jgeewax** [comment]
Some references.... (Short answer... not really. The best they have is `sudo apt-get install linuxbrew && sudo brew install grpc`)

- https://github.com/grpc/grpc/issues/936
- https://github.com/grpc/grpc/issues/1165

I believe the issue here is that gRPC requires some crazy stuff in SSL libs and others, which won't make it into mainstream world for quite some time, and users can't have two versions of those libraries the way apt handles packaging (for Linux).

As for @carterpage 's comment...

> We're designed to be very low latency and high-throughput so all of our focus is on gRPC. There is a REST interface, which comes for free with our envelope, but we don't recommend using for much more than slow admin API calls or ad hoc test queries against the data API.

HBase today doesn't speak HTTP 2.0 right? So wouldn't the REST-ful HTTP API calls be comparable to existing HBase clusters in latency and HTTP overhead?

**Comment by carterpage** [comment]
HBase has their own bespoke RPC system that's actually fairly fast.  The issue for us is, using Google's plumbing, gRPC is the only option to provide the sort of latencies we need.

RESTful HTTP API calls would be better than running a Thrift server in front of HBase, since that adds an extra hop, but that's a fairly low bar.  I think the performance comparison we want to aim for is Happybase against raw HBase.  I believe that uses HBase's binary RPC format, so I don't see how we'd match it without gRPC.

**Comment by jgeewax** [comment]
Got it. We won't be able to "depend" on this, so that means in our docs we'll basically have to say "If you want to use Cloud Bigtable via gcloud-python, you'll need to follow these steps to install gRPC on your machine".

This will clearly have more drop-off, but I suspect we already have a somewhat small pool of people using Cloud Bigtable (given the lowest cost product we offer people is $1500 / month), so adding an extra "here install this" hurdle wouldn't be the end of the world.

**Comment by dhermes** [comment]
Shouldn't the dev exp be even better for the people paying more? (Sorry if that is an unpopular opinion.)

**Comment by jgeewax** [comment]
Haha - yes, it should. The developer experience in this case is blazing fast managed HBase -- which is pretty incredible (and what you're paying for).

As far as "installing the client", it seems that our problem with packaging gRPC is not really in our control. The library itself (I'm told) depends on some pretty serious changes to libraries which are in the pipeline but not aimed for main-stream for quite some time. The best we can do (again, I'm told) is using linux brew (`brew install grpc`).

**Comment by carterpage** [comment]
JJ hit the nail on the head.  This is a pretty bleeding edge service, so there are still some rough edges that will get polished as more network stacks are comfortable with HTTP2, etc..  Our typical customer is technically sophisticated and willing to tolerate a little extra effort for the added power.  Hopefully we'll have the best of both worlds before long.

**Comment by dhermes** [comment]
So you're extending existing Python libraries to support HTTP2 or making your own? Why not just ship a custom fork?

Is the issue compiling C extensions for Python or is the issue making new versions of system socket libraries (e.g. things in `/usr/lib`)? It should still be possible to just point to local paths for things that would typically be found in `/usr/lib`.

**Comment by dhermes** [comment]
Some context for Python 3 support in `happybase` (and deps):

https://github.com/wbolster/happybase/issues/40
https://github.com/wbolster/happybase/pull/78
https://github.com/wbolster/happybase/pull/95

**Comment by maxluebbe** [comment]
Let's not think about Python 3 at this time. Getting a solid Python 2.x experience is our top priority.

**Comment by tseaver** [comment]
@maxluebbe the rest of `gcloud-python` is already Python3-compatible.  At a minimum, we need all "hard" dependencies to be compatible as well, so that our CI story stays sane.  I guess we could keep our code source-compatible, make `happylib` a "soft" dependency, and define the wrappers / tests only if it is importable.

**Comment by dhermes** [comment]
@jgeewax Suggested we use `happybase`. I'm trying to get a sense of its popularity after @maxluebbe made me step back and ask "why happybase?".

It only has [17 tagged][1] StackOverflow questions (as of June 10 at noon Pacific).

Checking installs is not so small:

```bash
$ pip install -U vanity
$ vanity happybase
...
happybase has been downloaded 487933 times!
```

This is in comparsion to 37,602,602 for `boto` and 10,308,673 for `httplib2`, so I'm not sure how to gauge it.

[1]: http://stackoverflow.com/questions/tagged/happybase

**Comment by jgeewax** [comment]
Well keep in mind that HBase is far less commonly used... The price point alone would be somewhat prohibitive for people here... (cheapest thing Cloud Bigtable sells is $1500 / month).

Think of it a bit like RedShift, costing $13k/TB/year -- if you could download a RedShift only library I wouldn't expect many installs showing up in vanity...

**Comment by jgeewax** [comment]
Also -- about the Python2/3 situation:

We're already effectively saying that you need to manually install gRPC. Would you guys prefer to have a separate library to install that is Python 2 only and requires gRPC? or to make a submodule that we flag as "won't run under Python 3, sorry" ?

/cc @maxluebbe

**Comment by dhermes** [comment]
Yeah the separate library (or component) (e.g. `pip install ipython[all]`) is likely a good way to go.

@jgeewax Is there some reason you feel like `happybase` is the "one true" library for HBase? Some recommendation from somewhere or any other quantifiable signal?

**Comment by maxluebbe** [comment]
@dhermes We've had requests from trusted testers to make this library work, and were directed by product management to make this work. Is there an alternative Python HBase client library you have in mind?

**Comment by dhermes** [comment]
None at all. I am a total n00b here, just wanted to know where it came from.

We'd love to hear from the TTs here too.

FWIW `happybase` seems to be a fine library and has many very similar concepts to this one. I just have no basis to understand its prevalence.

----

Also @tseaver mentioned previously that we may end up / be better off just making

```python
import happybase_gcloud as happybase
```

work without any other code changes necessary.

**Comment by jgeewax** [comment]
@dhermes : Any status updates here? Can we maybe come up with the list of crap we'd need to do?

**Comment by dhermes** [comment]
Sorry was boarding a plane when I saw this comment :) Working on it today

**Comment by dhermes** [comment]
I've been hacking on this, but `grpc` install is still posing a big issue.

Which [protos][1] need to be compiled for the service to work? It doesn't seem like there are **any** docs (on https://cloud.google.com/bigtable/docs/) for the actual web API, but maybe I'm supposed to use the `.proto` files as documentation. Every doc just assumes you're using an HBase client, though the python samples ([rest][2] and [thrift][3] may also provide some "docs" (a bit cowboy-coding-esque).

[1]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/tree/master/bigtable-protos/src/main/proto/google/bigtable/v1approved
[2]: https://github.com/GoogleCloudPlatform/cloud-bigtable-examples/tree/master/python/rest
[3]: https://github.com/GoogleCloudPlatform/cloud-bigtable-examples/tree/master/python/thrift

**Comment by lesv** [comment]
The [HappyBase API](https://happybase.readthedocs.org/en/latest/) would be the ideal API for Python, based on the [HBase API](https://hbase.apache.org/apidocs/) we use for Java [client](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/).  HappyBase talks to a Thrift server, but we'd rather it talked directly to a Cloud Bigtable server.

For what you are doing, you might benefit by looking at the [GoLang](https://godoc.org/google.golang.org/cloud/bigtable) [Client](https://github.com/GoogleCloudPlatform/gcloud-golang/tree/master/bigtable) as it's use of the [gRPC protos for Bigtable](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/tree/master/bigtable-protos) might help.

**Comment by dhermes** [comment]
@lesv Sadly there is a custom gRPC Go client, so its much easier for them to integrate. The install story for the gRPC C/C++ libraries is the biggest hang-up. As for the protos, I already linked to them and as for HappyBase, that is the current plan.

**Comment by lesv** [comment]
The protos I from the HBase Client repo are the ones you need:  [gRPC protos for Bigtable](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/tree/master/bigtable-protos)

**Comment by jgeewax** [comment]
@dhermes : Any progress to report ? People getting itchy about this one........

**Comment by dhermes** [comment]
I still haven't gotten what I [needed][1].

Main blockers:

1. `grpc` install doesn't really work
1. Which [protos][4] need to be compiled for the service to work? (Note, my last link was broken in the last 12 days because it was renamed from `v1approved` to `v1`.)

If I want to put `grpc` aside for the moment and just focus on the Cloud BigTable server (@lesv says "HappyBase talks to a Thrift server, but we'd rather it talked directly to a Cloud Bigtable server.")

1. There are no docs (on https://cloud.google.com/bigtable/docs/) for the actual web API
1. (Relating to above) Am I supposed to use the `.proto` files as documentation? Cowboy coding may work for Xooglers and really motivated OSS contributors, but do we expect users to do the same?
1. Should I use the python samples ([rest][2] and [thrift][3]) as my "docs" for the server? (The [`rest_client.py`][5] certainly seems to show some REST endpoints, but the README for it says "HBase REST Gateway setup and configuration" so it's probably just thrift over REST.)

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/issues/872#issuecomment-115051390
[2]: https://github.com/GoogleCloudPlatform/cloud-bigtable-examples/tree/master/python/rest
[3]: https://github.com/GoogleCloudPlatform/cloud-bigtable-examples/tree/master/python/thrift
[4]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/tree/1636e1c52057d94b823c1a7c9ddf4faaf96fb7c7/bigtable-protos/src/main/proto/google/bigtable/v1
[5]: https://github.com/GoogleCloudPlatform/cloud-bigtable-examples/blob/55c9da4ad6852ef32dc480a6b2ec726d5544ed2b/python/rest/rest_client.py

**Comment by jgeewax** [comment]
OK @dhermes, can you list out the specific things you need here so we can start getting them for you ?

**Comment by dhermes** [comment]
If we want to go directly to the Cloud BigTable server:

1. Which `.proto` files define the insert/retrieve endpoints? (I.e. typically REST-y endpoints.)
1. What `protoc` command should be run? Do i need the proto3 compiler (needed for `grpc`)? Is this documented anywhere?
1. Can I ship compiled Python modules (compiled from `.proto` files) with the library or does every user need to compile them on their own? (I hope the former.)
1. IIUC, gRPC talks to (mostly) the same endpoints, but over HTTP/2 instead of HTTP/1.1. Where are the docs for this gRPC API? Or am I just supposed to use the protos with nothing else documented?

**Comment by jgeewax** [comment]
Re 3: I'm being told by @lesv that Proto3 won't compile Python that can run on different systems... I'm astounded at this, but apparently it's true...

Can anyone double confirm as this seems like it'd be a showstopper... ?

**Comment by dhermes** [comment]
I'd guess the Python files are the same, but the actual native dependencies are not. It could also be possible that the compiled Python files are just `.so` C extensions, which would be platform dependent.

**Comment by lesv** [comment]
In looking over the [grpc.io](https://grpc.io) docs, it looks like the .so files are only required for building the thing and for the http2 stuff.  The protos are in python.  (sorry for the confusion)

**Comment by jgeewax** [comment]
OK - can someone build us some Python from those protos? :)

**Comment by jgeewax** [comment]
Nevermind -- just did it, sent the protos to @dhermes via e-mail for him to include in his PR.

**Comment by dhermes** [comment]
Made https://github.com/dhermes/gcloud-python-bigtable and will ping this thread when there is something working. Eventually we can prune the commit history and move the repo over to `GoogleCloudPlatform` (or integrate it into this repo).

**Comment by dhermes** [comment]
After a false start using HTTP/1.1 (not supported with the BigTable API) I've successfully got the `_pb2` modules [built][1] with the gRPC extensions.

However the basic docs don't explain how to call an external service:
http://www.grpc.io/docs/installation/python.html
http://www.grpc.io/docs/tutorials/basic/python.html
(e.g. how to set the path for a request, how to authenticate, etc.)

Would @nathanielmanistaatgoogle or @tbetbetbe or someone else on the gRPC team be willing to chat?

Right now I'm trying to adapt the sample code via:
https://gist.github.com/dhermes/2edb97d9581b5ec471eb
and am not having success (can't tell if a request is succeeding or failing).

[1]: https://github.com/dhermes/gcloud-python-bigtable/blob/8c4bc2261aa96c52936164924ccd319eb00cdf07/Makefile#L12-L69

**Comment by tbetbetbe** [comment]
@nathanielmanistaatgoogle will be back from holiday next week, and will follow-up on my comments then.

At the moment, the python auth example is missing, there's a PR to get that fixed
In the meantime, please see how auth + secure access is done in the [python interop test client](https://github.com/grpc/grpc/blob/master/src/python/interop/interop/client.py)

In particular, see
* [_oauth_access_token](https://github.com/grpc/grpc/blob/master/src/python/interop/interop/client.py#L66)
* [_stub](https://github.com/grpc/grpc/blob/master/src/python/interop/interop/client.py#L71)

Notes
* in _oauth_access_token, use the appropriate scopes.
* the SSL certificates to use are the prod ones, as loaded here (https://github.com/grpc/grpc/blob/master/src/python/interop/interop/client.py#L80}

**Comment by dhermes** [comment]
Thanks @tbetbetbe

**Comment by dhermes** [comment]
FYI, prod certificates [require][1] the `SSL_CERT_FILE` environment variable to be set (which isn't done by default on Linux AFAICT), so I'm not sure how that works for the typical user. I [adapted][2] a ruby checker for `SSL_CERT_FILE` and ran it on my Ubuntu 14.04 install and a default GCE instance. Both find the `SSL_CERT_FILE` at `/etc/ssl/certs/ca-certificates.crt`.

[1]: https://github.com/grpc/grpc/blob/a9a8eb70604b5da61f8e768bc52192910ba60251/src/python/interop/interop/resources.py#L46-L47
[2]: https://gist.github.com/dhermes/bbc5b7be1932bfffae77

**Comment by dhermes** [comment]
@tbetbetbe Just got a chance to implement the auth stuff (in my [gist][1]) and still getting a request timeout.

@nathanielmanistaatgoogle Can we get together this week?

[1]: https://gist.github.com/dhermes/2edb97d9581b5ec471eb

**Comment by dhermes** [comment]
Status update: Meeting with Nathaniel today. Hopefully we can iron out issues with request failures.

**Comment by dhermes** [comment]
I declare victory! Just leaving GOOG SFO after spending 3.5 hours with @nathanielmanistaatgoogle

We were banging our head against the wall for awhile and before giving up, fired up the go client (docs there could be better).

Eventually, we realized two things

- I stupidly never activated the API service
- the cluster admin API is **not public**

I was trying to do the simplest thing, ListZones with the cluster API, but it was not possible since the API is private (though the photos are released publicly)

In addition, the go error messages (from the server) were much more useful than those from gRPC in Python (not sure why).

I am headed to gym right now but look forward to getting this working soon.

**Comment by lesv** [comment]
The cluster admin API is public -- but it needs to be enabled separately.

**Comment by dhermes** [comment]
Sorry, I should have said that I couldn't enable it from the APIs console from a public account. Beyond that I don't know.

@lesv how does one enable the API for s project?

**Comment by lesv** [comment]
cloud.google.com/console
select your project, then select APIs & Auth > APIs and type in 'bigtable', you should see two apis listed, you need to enable both of them.

**Comment by dhermes** [comment]
Yes I know, they are the Bootable API (just for data within an existing table or tables) and the Table Admin API (for creating and deleting tables).

There is a 3rd API, the **Cluster** Admin API (I said "ListZones with the cluster API" above). This is the one I said isn't public.

**Comment by lesv** [comment]
Odd, I'm able to use the gcloud alpha bigtable clusters list & gcloud alpha bigtable clusters create from my personal account, with only the two APIs enabled on a project.  Both of them work and it's written in Python.

You can view the source on your machine google-cloud-sdk/lib/googlecloudsdk/bigtable/commands/clusters/*  it's python code.

**Comment by bendemaree** [comment]
My understanding with the Cloud Platform APIs that are alpha is that the project has to be whitelisted to access them (or even enable them). For example, alpha commands in the Compute Engine API are not available and I don't have an option to enable them.

**Comment by lesv** [comment]
The API's are ALPHA and public / published.  Try doing `gcloud components update` followed by `gcloud alpha bigtable clusters list --project <projectID>`  -- if it works, you'll have your answer.

I did this on my personal account and got good results.  That doesn't appear to be accurate for this, but you may need to: `gcloud components list`, and `gcloud components update alpha` first.

**Comment by dhermes** [comment]
@lesv I was currently in the process of this with `--log-http`. That doesn't mean `ListZones` will work, but `ListClusters` works and that is under the cluster admin API. I also noticed that `?alt=json` is appended to the URI.

It'd be great if someone would document the API surface, because right now I'm feeling around in the dark. I had to go through the source for the [scopes][5] and the service [endpoints][6].

There are clearly 3 services:

- [BigTable][1]
- [BigTable Cluster Admin][2]
- [BigTable Table Admin][3]

and the two APIs that can be enabled **seem** to correspond to only two of those three (to an outsider, which is all users)

![screen_shot_004](https://cloud.githubusercontent.com/assets/520669/8712802/b7b69020-2b15-11e5-80f2-e9efc7ea66a2.png)

Another distinction may be a user account vs. a service account.

Yet another may be the scopes used. [Checking][4] the access token (which I got from `--log-http`), none of the BigTable scopes are used

```
https://www.googleapis.com/auth/appengine.admin
https://www.googleapis.com/auth/bigquery
https://www.googleapis.com/auth/compute
https://www.googleapis.com/auth/devstorage.full_control
https://www.googleapis.com/auth/userinfo.email
https://www.googleapis.com/auth/ndev.cloudman
https://www.googleapis.com/auth/cloud-platform
https://www.googleapis.com/auth/sqlservice.admin
https://www.googleapis.com/auth/prediction
https://www.googleapis.com/auth/projecthosting
https://www.googleapis.com/auth/plus.me
```

but the catch-all `https://www.googleapis.com/auth/cloud-platform` is.

----

I'll report back with what I find about adding `?alt=json` and similar

[1]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/3ddf0189c0a18c0a918e9265fe892ca55e168264/bigtable-protos/src/main/proto/google/bigtable/v1/bigtable_service.proto
[2]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/3ddf0189c0a18c0a918e9265fe892ca55e168264/bigtable-protos/src/main/proto/google/bigtable/admin/cluster/v1/bigtable_cluster_service.proto
[3]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/3ddf0189c0a18c0a918e9265fe892ca55e168264/bigtable-protos/src/main/proto/google/bigtable/admin/table/v1/bigtable_table_service.proto
[4]: https://developers.google.com/apis-explorer/#p/oauth2/v2/oauth2.tokeninfo
[5]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/3ddf0189c0a18c0a918e9265fe892ca55e168264/bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialFactory.java#L42-L53
[6]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/3ddf0189c0a18c0a918e9265fe892ca55e168264/pom.xml#L73-L75

**Comment by lesv** [comment]
I sent a note to engineering earlier when you said it doesn't work.  I seem to recall that both admin API's are enabled by a single flag for simplicity, but I'm confirming it.  The bigtable scopes work, but cloud-platform is preferred -- we wanted to simplify things.  But it also makes it asymmetric for a while which confuses.

**Comment by bendemaree** [comment]
@lesv You were correct about the `clusters list` request; thanks!

**Comment by dhermes** [comment]
@lesv TLDR; It seems the issue is using a service account rather than a user account.

----

I set some environment variables for config

```bash
PROJECT_ID="some-id-from-console"
BEARER_TOKEN="oauth2-token"  # Can get user token via --log-http output
```

----

With the same project ID I made two different cluster admin requests (`ListClusters` and `ListZones`) using three different tokens

-   The user token minted via `gcloud auth login` (for my personal account)
-   A service account for the project, with the scopes 

    ```
    https://www.googleapis.com/auth/cloud-bigtable.data
    https://www.googleapis.com/auth/cloud-bigtable.admin
    ```
-   A service account for the project (the same one from 2), with the scope

    ```
    https://www.googleapis.com/auth/cloud-platform
    ```

----

To send the requests, I did as follows (using the env. vars from above):

```bash
$ # ListClusters
$ curl \
> --header "Authorization: Bearer ${BEARER_TOKEN}" \
> --header "accept: application/json" \
> --header "accept-encoding: gzip, deflate" \
> --header "content-length: 0" \
> --header "user-agent: dhermes-curl-bigtable-test" \
> https://bigtableclusteradmin.googleapis.com/v1/projects/${PROJECT_ID}/aggregated/clusters?alt=json
$
$ # ListZones
$ curl \
> --header "Authorization: Bearer ${BEARER_TOKEN}" \
> --header "accept: application/json" \
> --header "accept-encoding: gzip, deflate" \
> --header "content-length: 0" \
> --header "user-agent: dhermes-curl-bigtable-test" \
> https://bigtableclusteradmin.googleapis.com/v1/projects/${PROJECT_ID}/zones?alt=json
```

----

Using the user account the response for ListClusters was

```json
{}
```

and for ListZones

```json
{
  "zones": [
    {
      "name": "projects/REDACTED/zones/asia-east1-b",
      "displayName": "asia-east1-b",
      "status": "OK"
    },
    {
      "name": "projects/REDACTED/zones/europe-west1-c",
      "displayName": "europe-west1-c",
      "status": "OK"
    },
    {
      "name": "projects/REDACTED/zones/us-central1-c",
      "displayName": "us-central1-c",
      "status": "OK"
    },
    {
      "name": "projects/REDACTED/zones/us-central1-b",
      "displayName": "us-central1-b",
      "status": "OK"
    }
  ]
}
```

----

On the other hand both methods failed for the service account (no matter which scopes were used) with the message

```json
{
  "error": {
    "code": 403,
    "message": "Project has not enabled the API. Please use Google Developers Console to activate the API for your project.",
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.Help",
        "links": [
          {
            "description": "Google developer console API activation",
            "url": "https://console.developers.google.com/project/REDACTED/apiui/api"
          }
        ]
      }
    ]
  }
}
```

----

We **have to** encourage a service account to be used for deployed applications. That is the point of a service account. So these endpoints need to work for service accounts.

**Comment by lesv** [comment]
Agree - it shouldn't work the way you describe.  I'm not even sure where to send this, but I'll mention it to the eng team.  @jgeewax may have some insight.

**Comment by jgeewax** [comment]
@carterpage @coryoconnor -- Any ideas on why service accounts are no good?

**Comment by dhermes** [comment]
@jgeewax Small distinction: service accounts work (i.e. they are "good") for the Table Admin API and Data API, just not the Cluster Admin API (here is where they are "no good").

**Comment by coryoconnor** [comment]
@dmmcerlean @sduskis any ideas?

**Comment by lesv** [comment]
Try with one of the following scopes:
https://www.googleapis.com/auth/cloud-bigtable.admin
https://www.googleapis.com/auth/cloud-bigtable.admin.cluster

**Comment by dhermes** [comment]
Trying only 1 at a time as wells as both at once, it still fails with the

> Project has not enabled the API. Please use Google Developers Console to activate the API for your project.

----

If you like, I've made an all-inclusive gist that makes this easy to run and tweak:
https://gist.github.com/dhermes/d27070c90a9862213a3b

The script `main_with_cluster_admin.go` [uses][1] `bigtable.ClusterAdminScope`. This value ([defined][2] in `doc.go`) is slightly different than the one you suggested:

```
https://www.googleapis.com/auth/bigtable.admin.cluster        # In GOLANG
https://www.googleapis.com/auth/cloud-bigtable.admin.cluster  # Your suggestion
```

[1]: https://gist.github.com/dhermes/d27070c90a9862213a3b#file-main_with_cluster_admin-go-L11
[2]: https://github.com/GoogleCloudPlatform/gcloud-golang/blob/cb92b5f0a2b35eb37a887bb0821fd4d67ebe4bf6/bigtable/doc.go#L103

**Comment by dhermes** [comment]
Here is a well-polished jumping off point for Python:
https://gist.github.com/dhermes/2edb97d9581b5ec471eb

Will merge this upstream into https://github.com/dhermes/gcloud-python-bigtable

**Comment by lesv** [comment]
@AngusDavis -- Please take a look.

**Comment by lesv** [comment]
I've talked with engineering -- it is probably best not to provide the cluster APIs for now.

**Comment by jgeewax** [comment]
That doesn't quite make sense. Why not?

**Comment by lesv** [comment]
gcloud, when acquiring OAuth tokens, use the client id and client secret (from the installed application OAuth flow). This results in gcloud being able to use the API, but not tokens associated with service accounts.

When the Cluster API is fully enabled, it will be usable programmatically. Until then, the recommendation is to use gcloud or cloud console for the infrequent case of creating a cluster and the data and table APIs for the more frequent cases.

So basically, the API is visible to some internal projects and not to everyone else.  We're trying to get this exposed via a service acct, it's currently not... hang tight

**Comment by dhermes** [comment]
Hanging tight; from the outside perspective, service accounts should be prioritized over user accounts since service accounts are used in a deployed application.

**Comment by dhermes** [comment]
OK here is a new one, I can't even create a cluster through the API anymore:

![screen_shot_006](https://cloud.githubusercontent.com/assets/520669/8783143/d34155a2-2ece-11e5-9519-e5c4b2e35d61.png)

Note the "Table Admin API" service is enabled on the project

![screen_shot_007](https://cloud.githubusercontent.com/assets/520669/8783161/ee1df614-2ece-11e5-8770-f0c553d5b453.png)

and as mentioned before there is no Cluster Admin API as a choice to enable:

![screen_shot_008](https://cloud.githubusercontent.com/assets/520669/8783170/01dddc00-2ecf-11e5-9bab-9db0debf5082.png)

**Comment by lesv** [comment]
I've reported it to engineering.

**Comment by dhermes** [comment]
Thanks. I think it's just my personal (@gmail.com) account. @jgeewax was able to create one through the UI with a @google.com account.

**Comment by lesv** [comment]
Yep -- it fails on my personal gmail account when I'm not on the corporate network as well.

**Comment by lesv** [comment]
Fix is rolling out now, should be done in the next few hours.

**Comment by dhermes** [comment]
Fix for UI or fix for supporting service accounts with Cluster Admin API?

**Comment by lesv** [comment]
At this point, it should be just like things were last week.  We rolled back some changes.

**Comment by dhermes** [comment]
Gotcha. Thanks.

**Comment by dhermes** [comment]
@lesv ISTM that it became possible to use service accounts for the Cluster Admin API yesterday. Can you confirm? (Though the API still doesn't show up in the list of APIs to enable on cloud console.)

**Comment by lesv** [comment]
That might have happened as part of the fix for the problem you found recently -- the engineer I need to talk w/ is out this week -- so, don't take it as a given that it will stay this way.  Eventually, it will be correct and you can use them -- but I'm not sure it has happened officially yet.

**Comment by lesv** [comment]
@dhermes @jgeewax The Cluster Admin API is definitely not supposed to be enabled yet.  The cluster proto was open sourced for some testing reasons, but wasn't supposed to be made available externally yet.  One reason is that there are some naming issues that have come up, and the proto may have to be changed because of this.  That conversation won't take place until next week and we probably won't see any public results for at least a week after that.

**Comment by dhermes** [comment]
@lesv Thanks. I have the entire Cluster Admin API [implemented][1], including

- an automated generator for the `*_pb2.py` modules
- a system test suite that covers every API method (except for `UndeleteCluster`) with real API calls
- a unit test suite with 100% code coverage

----

Though it's not supposed to be enabled yet, any changes will be quickly picked up by the 3 suites mentioned above.

Thanks for the heads up though.

----

PS ISTM that supporting `cloud alpha bigtable clusters *` would be impossible without enabling the API.

[1]: https://github.com/dhermes/gcloud-python-bigtable

**Comment by dhermes** [comment]
I have everything implemented except for the stream requests (`ReadRows` and `SampleRowKeys`), which I'm not sure how to handle with gRPC Python.

Golang handles it by [using][1] a cancel-able stream.

[1]: https://github.com/GoogleCloudPlatform/gcloud-golang/blob/e34a32f9b0ecbc0784865fb2d47f3818c09521d4/bigtable/bigtable.go#L105

**Comment by nathanielmanistaatgoogle** [comment]
@dhermes: what's unclear? gRPC Python supports cancellation. (Ping me; I have some time open this afternoon if you'd like to videochat.)

**Comment by dhermes** [comment]
Nothing unclear (that I know of), I just don't know how to do it. Let's chat.

**Comment by jgeewax** [comment]
@dhermes : Looks like we're "done enough" with the Bigtable stuff. I'm going to close this out and open a new issue focused on "what exactly do we need to do to make `gcloud_bigtable` become `gcloud.bigtable`"

Sound reasonable? (Re-open if not)
